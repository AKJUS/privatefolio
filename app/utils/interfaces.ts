export interface Trade extends Record<string, number | string> {
  amount: number;
  datetime: string;
  executedAmount: string;
  fee: string;
  filledPrice: number;
  id: number;
  role: "Maker" | "Taker";
  side: "BUY" | "SELL";
  symbol: string;
  ticker: string;
  total: number;
}
