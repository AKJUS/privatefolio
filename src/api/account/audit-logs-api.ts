/* eslint-disable sort-keys-fix/sort-keys-fix */
import { proxy } from "comlink"
import { noop } from "src/utils/utils"

import { AuditLog } from "../../interfaces"
import { ProgressCallback } from "../../stores/task-store"
import { getAccount } from "../database"

const _filterOrder: (keyof AuditLog)[] = ["platform", "wallet", "operation", "assetId"]
const _filterOrderBySpecificity: (keyof AuditLog)[] = ["assetId", "operation", "wallet", "platform"]

export async function indexAuditLogs(progress: ProgressCallback = noop, accountName: string) {
  const account = getAccount(accountName)
  progress([0, "Audit logs: cleaning up stale indexes"])
  await account.auditLogsDB.viewCleanup()
  progress([10, "Audit logs: updating index for 'timestamp'"])
  await account.auditLogsDB.createIndex({
    index: {
      fields: ["timestamp"],
      name: "timestamp",
    },
  })
  progress([20, "Audit logs: updating index for 'platform'"])
  await account.auditLogsDB.createIndex({
    index: {
      fields: ["platform", "timestamp", "wallet", "operation", "assetId"], // MUST respect the order in _filterOrder
      name: "platform",
    },
  })
  progress([30, "Audit logs: updating index for 'wallet'"])
  await account.auditLogsDB.createIndex({
    index: {
      fields: ["wallet", "timestamp", "platform", "operation", "assetId"], // MUST respect the order in _filterOrder
      name: "wallet",
    },
  })
  progress([40, "Audit logs: updating index for 'operation'"])
  await account.auditLogsDB.createIndex({
    index: {
      fields: ["operation", "timestamp", "platform", "wallet", "assetId"], // MUST respect the order in _filterOrder
      name: "operation",
    },
  })
  progress([45, "Audit logs: updating index for 'assetId'"])
  await account.auditLogsDB.createIndex({
    index: {
      fields: ["assetId", "timestamp", "platform", "wallet", "operation"], // MUST respect the order in _filterOrder
      name: "assetId",
    },
  })
  progress([50, "Audit logs: updating index for 'txId'"])
  await account.auditLogsDB.createIndex({
    index: {
      fields: ["txId"],
      name: "txId",
    },
  })
}

type FindAuditLogsRequest = {
  fields?: string[]
  filters?: Partial<Record<keyof AuditLog, string | number | PouchDB.Find.ConditionOperators>>
  limit?: number
  /**
   * orderBy = timestamp, always
   *
   * @default "desc"
   */
  order?: "asc" | "desc"
  skip?: number
}

export async function findAuditLogs(request: FindAuditLogsRequest = {}, accountName: string) {
  const account = getAccount(accountName)
  const { indexes } = await account.auditLogsDB.getIndexes()
  if (indexes.length === 1) {
    // await indexAuditLogs(console.log, accountName)
    await indexAuditLogs(undefined, accountName)
  }

  const { filters = {}, limit, skip, order = "desc", fields } = request

  // Algorithm to help PouchDB find the best index to use
  const preferredFilter = _filterOrderBySpecificity.find((x) => filters[x])

  const selector: PouchDB.Find.Selector = !preferredFilter
    ? { timestamp: { $exists: true } }
    : {
        [preferredFilter]: filters[preferredFilter],
        timestamp: { $exists: true },
      }

  if (preferredFilter) {
    _filterOrder.forEach((filter) => {
      if (filter === preferredFilter) return
      selector[filter] = filters[filter] ? filters[filter] : { $exists: true }
    })
  }

  const sort: PouchDB.Find.FindRequest<AuditLog>["sort"] = !preferredFilter
    ? [{ timestamp: order }]
    : [{ [preferredFilter]: order }, { timestamp: order }]

  const _req: PouchDB.Find.FindRequest<AuditLog> = {
    fields,
    limit,
    selector,
    skip,
    sort,
  }
  // console.log("ðŸ“œ LOG > findAuditLogs > _req:", _req)
  // const explain = await (account.auditLogsDB as any).explain(_req)
  // console.log("ðŸ“œ LOG > findAuditLogs > explain:", explain.index)

  //
  const { docs, warning } = await account.auditLogsDB.find(_req)
  if (warning) console.warn("findAuditLogs", warning)
  return docs as AuditLog[]
}

export async function findAuditLogsForTxId(txId: string, accountName: string) {
  const account = getAccount(accountName)
  const { indexes } = await account.auditLogsDB.getIndexes()
  if (indexes.length === 1) {
    // await indexAuditLogs(console.log, accountName)
    await indexAuditLogs(undefined, accountName)
  }

  const _req: PouchDB.Find.FindRequest<AuditLog> = {
    selector: {
      txId,
    },
  }
  // console.log("ðŸ“œ LOG > findAuditLogsForTxId > _req:", _req)
  // const explain = await (account.auditLogsDB as any).explain(_req)
  // console.log("ðŸ“œ LOG > findAuditLogsForTxId > explain:", explain.index)

  //
  const { docs, warning } = await account.auditLogsDB.find(_req)
  if (warning) console.warn("findAuditLogsForTxId", warning)
  return docs as AuditLog[]
}

export async function countAuditLogs(accountName: string) {
  const account = getAccount(accountName)
  const indexes = await account.auditLogsDB.allDocs({
    include_docs: false,
    // Prefix search
    // https://pouchdb.com/api.html#batch_fetch
    endkey: `_design\ufff0`,
    startkey: "_design",
  })
  const result = await account.auditLogsDB.allDocs({ include_docs: false, limit: 1 })
  return result.total_rows - indexes.rows.length
}

export function subscribeToAuditLogs(callback: () => void, accountName: string) {
  const account = getAccount(accountName)
  const changesSub = account.auditLogsDB
    .changes({
      live: true,
      since: "now",
    })
    .on("change", callback)

  return proxy(() => {
    try {
      changesSub.cancel()
    } catch {}
  })
}
