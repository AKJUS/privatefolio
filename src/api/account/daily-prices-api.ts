import { proxy } from "comlink"
import { AssetId, ChartData, ResolutionString, SavedPrice, Time, Timestamp } from "src/interfaces"
import { PRICE_API_PAGINATION, PRICE_APIS_META, PriceApiId } from "src/settings"
import { ProgressCallback } from "src/stores/task-store"
import { getAssetTicker } from "src/utils/assets-utils"
import { formatDate } from "src/utils/formatting-utils"
import { noop } from "src/utils/utils"

import { getAccount } from "../database"
import { validateOperation } from "../database-utils"
import { PAIR_MAPPER, PRICE_APIS, PRICE_MAPPER } from "../external/prices/providers"
import { updateAsset } from "./assets-api"

export async function indexDailyPrices(accountName: string, progress: ProgressCallback = noop) {
  const account = getAccount(accountName)

  progress([0, "Daily prices: updating index for 'timestamp'"])
  await account.dailyPricesDB.createIndex({
    index: {
      fields: ["timestamp"],
      name: "timestamp",
    },
  })
  progress([50, "Daily prices: updating index for 'assetId'"])
  await account.dailyPricesDB.createIndex({
    index: {
      fields: ["assetId", "timestamp"],
      name: "assetId",
    },
  })
}

export async function getPricesForAsset(
  accountName: string,
  assetId: string,
  timestamp?: Timestamp
) {
  if (assetId === "USDT" && !!timestamp) {
    return [{ time: timestamp / 1000, value: 1 }] as ChartData[]
  }

  const account = getAccount(accountName)
  await indexDailyPrices(accountName)

  const _req: PouchDB.Find.FindRequest<SavedPrice> = {
    selector: {
      assetId,
      timestamp: timestamp || { $exists: true },
    },
    sort: [
      {
        //
        assetId: "asc",
        // eslint-disable-next-line sort-keys-fix/sort-keys-fix
        timestamp: "asc",
      },
    ],
  }

  // console.log("ðŸ“œ LOG > findTransactions > _req:", _req)
  // const explain = await (account.dailyPricesDB as any).explain(_req)
  // console.log("ðŸ“œ LOG > findTransactions > explain:", explain.index)

  const prices = await account.dailyPricesDB.find(_req)

  return prices.docs.map((x) => x.price)
}

export async function getAssetPriceMap(
  accountName: string,
  timestamp: Timestamp = new Date().getTime()
) {
  await indexDailyPrices(accountName)
  const account = getAccount(accountName)

  const day: Timestamp = timestamp - (timestamp % 86400000)

  const _req: PouchDB.Find.FindRequest<SavedPrice> = {
    fields: ["assetId", "price", "source"],
    selector: { timestamp: day },
  }

  // console.log("ðŸ“œ LOG > findTransactions > _req:", _req)
  // const explain = await (account.dailyPricesDB as any).explain(_req)
  // console.log("ðŸ“œ LOG > findTransactions > explain:", explain.index)

  const prices = await account.dailyPricesDB.find(_req)

  return prices.docs.reduce(
    (map, x) => {
      map[x.assetId] = x.price
      return map
    },
    {
      USDT: { time: (day / 1000) as Time, value: 1 },
    } as Record<string, ChartData>
  )
}

export async function getPriceCursor(
  accountName: string,
  assetId: string
): Promise<Timestamp | undefined> {
  const account = getAccount(accountName)
  const prices = await account.dailyPricesDB.find({
    limit: 1,
    selector: {
      assetId,
      timestamp: { $exists: true },
    },
    sort: [
      {
        assetId: "desc",
        timestamp: "desc",
      },
    ],
  })

  if (prices.docs.length === 0) {
    return undefined
  }

  return prices.docs[0].timestamp
}

type FetchDailyPricesRequest = {
  assetIds?: AssetId[]
  priceApiMap?: Partial<Record<AssetId, PriceApiId>>
}

export async function fetchDailyPrices(
  accountName: string,
  request: FetchDailyPricesRequest,
  progress: ProgressCallback = noop,
  signal?: AbortSignal
) {
  const { assetIds, priceApiMap = {} } = request

  if (!assetIds) {
    throw new Error("No assetIds provided") // TODO prevent this
  }

  progress([0, `Fetching asset prices for ${assetIds.length} assets`])
  await indexDailyPrices(accountName)
  const account = getAccount(accountName)

  const now = Date.now()
  const today: Timestamp = now - (now % 86400000)

  const promises: (() => Promise<void>)[] = []

  for (let i = 1; i <= assetIds.length; i++) {
    // eslint-disable-next-line no-loop-func
    promises.push(async () => {
      const assetId = assetIds[i - 1]

      if (signal?.aborted) throw new Error(signal.reason)

      const preferredPriceApiId = priceApiMap[assetId]
      const priceApiIds = preferredPriceApiId
        ? [preferredPriceApiId]
        : (Object.keys(PRICE_APIS) as PriceApiId[])

      let since: Timestamp | undefined = await getPriceCursor(accountName, assetId)
      let until: Timestamp | undefined = today

      if (!since) since = today - 86400000 * (PRICE_API_PAGINATION - 1)

      for (const priceApiId of priceApiIds) {
        try {
          while (true) {
            const priceApi = PRICE_APIS[priceApiId]
            const priceMapper = PRICE_MAPPER[priceApiId]
            const pairMapper = PAIR_MAPPER[priceApiId]

            if (!priceApi || !priceMapper || !pairMapper) {
              throw new Error(`Price API "${priceApiId}" is not supported`)
            }

            const pair = pairMapper(assetId)

            const results = await priceApi({
              limit: PRICE_API_PAGINATION,
              pair,
              since,
              timeInterval: "1d" as ResolutionString,
              until,
            })

            if (!preferredPriceApiId && results.length > 0) {
              updateAsset(accountName, assetId, { priceApiId })
            }

            if (signal?.aborted) {
              throw new Error(signal.reason)
            }

            if (results.length === 0) {
              throw new Error(`${PRICE_APIS_META[priceApiId].name} EmptyResponse`)
            }

            const docIds: Array<{ id: string }> = []

            const documentMap = results.reduce((acc, result) => {
              const price = priceMapper(result)
              const timestamp = (price.time as number) * 1000
              const _id = `${assetId}-${timestamp}`

              acc[_id] = {
                _id,
                assetId,
                pair,
                price,
                timestamp,
              }

              docIds.push({ id: _id })
              return acc
            }, {} as Record<string, SavedPrice>)

            const { results: docsWithRevision } = await account.dailyPricesDB.bulkGet({
              docs: docIds,
            })

            const docs: SavedPrice[] = docsWithRevision.map((x) => ({
              ...documentMap[x.id],
              _rev: "ok" in x.docs[0] ? x.docs[0].ok._rev : undefined,
            }))

            const updates = await account.dailyPricesDB.bulkDocs(docs)
            validateOperation(updates)

            const start = (priceMapper(results[0]).time as number) * 1000
            const end = (priceMapper(results[results.length - 1]).time as number) * 1000

            progress([
              undefined,
              `Fetched ${getAssetTicker(assetId)} using ${
                PRICE_APIS_META[priceApiId].name
              } from ${formatDate(start)} to ${formatDate(end)}`,
            ])

            if (results.length !== PRICE_API_PAGINATION) {
              // reached listing date (genesis)
              break
            }

            until = start - 86400000
            since = start - 86400000 * PRICE_API_PAGINATION
          }
          break
        } catch (error) {
          progress([undefined, `Skipped ${getAssetTicker(assetId)}: ${error}`])
        }
      }
    })
  }

  let progressCount = 0

  await Promise.all(
    promises.map((fetchFn) =>
      fetchFn().then(() => {
        progressCount += 1
        progress([(progressCount / assetIds.length) * 100])
      })
    )
  )
}

export function subscribeToDailyPrices(callback: () => void, accountName: string) {
  const account = getAccount(accountName)

  const changesSub = account.dailyPricesDB
    .changes({
      live: true,
      since: "now",
    })
    .on("change", callback)

  return proxy(() => {
    try {
      changesSub.cancel()
    } catch {}
  })
}
