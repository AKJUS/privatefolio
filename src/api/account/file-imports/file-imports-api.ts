import { proxy } from "comlink"
import { FileImport } from "src/interfaces"
import { ProgressCallback } from "src/stores/task-store"
import { hashString, noop } from "src/utils/utils"

import { getAccount } from "../../database"
import { parseCsv } from "./csv-utils"

export async function addFileImport(
  file: File,
  progress: ProgressCallback = noop,
  accountName = "main"
) {
  const account = getAccount(accountName)
  const { name, type, lastModified, size } = file

  if (type !== "text/csv") {
    throw new Error("Error reading file: not .csv")
  }

  const timestamp = new Date().getTime()
  const _id = hashString(`${name}_${size}_${lastModified}`)

  await account.fileImportsDB.put({
    _id,
    lastModified,
    name,
    size,
    timestamp,
  } as unknown as FileImport)

  // parse file
  const text = await file.text()
  const { metadata, logs, transactions } = await parseCsv(text, _id, progress)

  // save logs
  progress([60, `Saving ${logs.length} audit logs to disk`])
  await account.auditLogsDB.bulkDocs(logs)

  // save transactions
  progress([80, `Saving ${transactions.length} transactions to disk`])
  await account.transactionsDB.bulkDocs(transactions)

  // save metadata
  const fileImport = await account.fileImportsDB.get(_id)
  fileImport.meta = metadata
  await account.fileImportsDB.put<FileImport>(fileImport)

  return { _id, metadata }
}

export async function getFileImports(accountName = "main") {
  const account = getAccount(accountName)
  const res = await account.fileImportsDB.allDocs<FileImport>({
    include_docs: true,
  })
  return res.rows.map((row) => row.doc) as FileImport[]
}

export async function removeFileImport(
  fileImport: FileImport,
  progress: ProgressCallback,
  accountName = "main"
) {
  const account = getAccount(accountName)
  // TODO consider pagination
  const logs = await account.auditLogsDB.allDocs({
    // Prefix search
    // https://pouchdb.com/api.html#batch_fetch
    endkey: `${fileImport._id}\ufff0`,
    startkey: fileImport._id,
  } as PouchDB.Core.AllDocsWithinRangeOptions)
  progress([25, `Removing ${logs.rows.length} audit logs`])

  // await Promise.all(
  //   logs.rows.map((row) =>
  //     auditLogsDB.remove({
  //       _id: row.id,
  //       _rev: row.value.rev,
  //     })
  //   )
  // )

  await account.auditLogsDB.bulkDocs(
    logs.rows.map((row) => ({ _deleted: true, _id: row.id, _rev: row.value.rev } as never))
  )
  // Transactions
  const txns = await account.transactionsDB.allDocs({
    // Prefix search
    // https://pouchdb.com/api.html#batch_fetch
    endkey: `${fileImport._id}\ufff0`,
    startkey: fileImport._id,
  } as PouchDB.Core.AllDocsWithinRangeOptions)
  progress([50, `Removing ${txns.rows.length} transactions`])

  // await Promise.all(
  //   txns.rows.map((row) =>
  //     transactionsDB.remove({
  //       _id: row.id,
  //       _rev: row.value.rev,
  //     })
  //   )
  // )

  await account.transactionsDB.bulkDocs(
    txns.rows.map((row) => ({ _deleted: true, _id: row.id, _rev: row.value.rev } as never))
  )

  const res = await account.fileImportsDB.remove(fileImport)
  return res.ok
}

export function subscribeToFileImports(callback: () => void, accountName = "main") {
  const account = getAccount(accountName)
  const changesSub = account.fileImportsDB
    .changes({
      live: true,
      since: "now",
    })
    .on("change", callback)

  return proxy(() => {
    try {
      // console.log("ðŸ“œ LOG > subscribeToFileImports > unsubbing")
      changesSub.cancel()
    } catch {}
  })
}
