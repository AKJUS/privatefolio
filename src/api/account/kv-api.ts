import { proxy } from "comlink"

import { AccountDatabase, main } from "../database"

export async function setValue(key: string, value: unknown, account: AccountDatabase = main) {
  try {
    const existing = await account.keyValueDB.get(key)
    existing.value = value
    await account.keyValueDB.put(existing)
  } catch {
    await account.keyValueDB.put({ _id: key, value })
  }
}

export async function getValue<T>(
  key: string,
  defaultValue: T | null = null,
  account: AccountDatabase = main
): Promise<T | null> {
  try {
    const existing = await account.keyValueDB.get(key)
    return existing.value as T
  } catch {
    return defaultValue
  }
}

export function subscribeToKV(key: string, callback: () => void, account: AccountDatabase = main) {
  const changesSub = account.keyValueDB
    .changes({
      live: true,
      selector: {
        _id: key,
      },
      since: "now",
    })
    .on("change", callback)

  return proxy(() => {
    try {
      changesSub.cancel()
    } catch {}
  })
}
