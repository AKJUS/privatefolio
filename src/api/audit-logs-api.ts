import { AuditLog } from "../interfaces"
import { ActiveFilterMap } from "../stores/audit-log-store"
import { auditLogsDB } from "./database"

/**
 * Retrieves audit logs from the database.
 *
 * @warning Unsorted.
 */
export async function getAuditLogs(filters: ActiveFilterMap = {}) {
  await auditLogsDB.createIndex({
    index: {
      fields: ["symbol"],
    },
  })
  await auditLogsDB.createIndex({
    index: {
      fields: ["operation"],
    },
  })

  const res2 = await auditLogsDB.allDocs({ include_docs: false })
  console.log("ðŸ“œ LOG > getAuditLogs > res2:", res2)

  const result = await auditLogsDB.find({
    limit: 25,
    // attachments: true,
    // descending: true,
    // include_docs: true,
    selector: filters,
    sort: [{ symbol: "desc" }],
  })
  const { docs, warning } = result
  console.log("ðŸ“œ LOG > getAuditLogs > result:", result)

  if (warning) console.warn(warning)

  return docs as AuditLog[]
  // return res.rows.map((row) => row.doc)
}
