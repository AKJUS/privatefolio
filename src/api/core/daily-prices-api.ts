import {
  ChartData,
  PriceApiId,
  ResolutionString,
  SavedPrice,
  Time,
  Timestamp,
} from "src/interfaces"
import { PRICE_API_PAGINATION } from "src/settings"
import { ProgressCallback } from "src/stores/task-store"
import { formatDate } from "src/utils/formatting-utils"
import { noop } from "src/utils/utils"

import { core } from "../database"
import { PAIR_MAPPER, PRICE_APIS, PRICE_MAPPER } from "./prices/providers"

export async function indexDailyPrices() {
  await core.dailyPricesDB.createIndex({
    index: {
      fields: ["timestamp"],
      name: "timestamp",
    },
  })
  await core.dailyPricesDB.createIndex({
    index: {
      fields: ["symbol", "source", "timestamp"],
      name: "symbol",
    },
  })
}

export async function getPricesForAsset(
  symbol: string,
  source = "coinbase",
  timestamp?: Timestamp
) {
  if (symbol === "USDT" && !!timestamp) {
    return [{ time: timestamp / 1000, value: 1 }] as ChartData[]
  }

  await indexDailyPrices()

  const _req: PouchDB.Find.FindRequest<SavedPrice> = {
    selector: {
      symbol,
      // eslint-disable-next-line sort-keys-fix/sort-keys-fix
      source,
      timestamp: timestamp || { $exists: true },
    },
    sort: [
      {
        //
        symbol: "asc",
        // eslint-disable-next-line sort-keys-fix/sort-keys-fix
        source: "asc",
        timestamp: "asc",
      },
    ],
  }

  // console.log("ðŸ“œ LOG > findTransactions > _req:", _req)
  // const explain = await (core.dailyPricesDB as any).explain(_req)
  // console.log("ðŸ“œ LOG > findTransactions > explain:", explain.index)

  const prices = await core.dailyPricesDB.find(_req)

  return prices.docs.map((x) => x.price)
}

export async function getAssetPriceMap(timestamp: Timestamp) {
  await indexDailyPrices()

  const _req: PouchDB.Find.FindRequest<SavedPrice> = {
    fields: ["symbol", "price", "source"],
    selector: { timestamp },
  }

  // console.log("ðŸ“œ LOG > findTransactions > _req:", _req)
  // const explain = await (core.dailyPricesDB as any).explain(_req)
  // console.log("ðŸ“œ LOG > findTransactions > explain:", explain.index)

  const prices = await core.dailyPricesDB.find(_req)

  return prices.docs.reduce(
    (map, x) => {
      console.log("ðŸ“œ LOG > getAssetPriceMap > x:", x)
      map[x.symbol] = x.price
      return map
    },
    {
      USDT: { time: (timestamp / 1000) as Time, value: 1 },
    } as Record<string, ChartData>
  )
}

export async function getPriceCursor(
  symbol: string,
  source: PriceApiId
): Promise<Timestamp | undefined> {
  const prices = await core.dailyPricesDB.find({
    limit: 1,
    selector: {
      symbol,
      // eslint-disable-next-line sort-keys-fix/sort-keys-fix
      source,
      timestamp: { $exists: true },
    },
    sort: [
      {
        symbol: "desc",
        // eslint-disable-next-line sort-keys-fix/sort-keys-fix
        source: "desc",
        timestamp: "desc",
      },
    ],
  })

  if (prices.docs.length === 0) {
    return undefined
  }

  return prices.docs[0].timestamp
}

type FetchDailyPricesRequest = {
  apiPreference?: PriceApiId
  symbols?: string[]
}

export async function fetchDailyPrices(
  request: FetchDailyPricesRequest,
  progress: ProgressCallback = noop,
  signal?: AbortSignal
) {
  const { symbols, apiPreference = "coinbase" } = request

  if (!symbols) {
    throw new Error("No symbols provided") // TODO prevent this
  }

  progress([0, `Fetching asset prices for ${symbols.length} symbols`])
  await indexDailyPrices()

  const now = Date.now()
  const today: Timestamp = now - (now % 86400000)

  const priceApi = PRICE_APIS[apiPreference]
  const priceMapper = PRICE_MAPPER[apiPreference]
  const pairMapper = PAIR_MAPPER[apiPreference]

  if (!priceApi || !priceMapper || !pairMapper) {
    throw new Error(`Price api ${apiPreference} is not supported`)
  }

  for (let i = 1; i <= symbols.length; i++) {
    const symbol = symbols[i - 1]

    if (signal?.aborted) {
      throw new Error(signal.reason)
    }

    try {
      let since: Timestamp | undefined = await getPriceCursor(symbol, apiPreference)
      let until: Timestamp | undefined = today

      if (!since) since = today - 86400000 * PRICE_API_PAGINATION

      while (true) {
        const pair = pairMapper(symbol)
        const source = apiPreference

        const results = await priceApi({
          limit: PRICE_API_PAGINATION,
          pair,
          since,
          timeInterval: "1d" as ResolutionString,
          until,
        })

        const docIds: Array<{ id: string }> = []

        const documentMap = results.reduce((acc, result) => {
          const price = priceMapper(result)
          const timestamp = (price.time as number) * 1000
          const _id = `${pair}-${source}-${timestamp}`

          acc[_id] = {
            _id,
            pair,
            price,
            source,
            symbol,
            timestamp,
          }

          docIds.push({ id: _id })
          return acc
        }, {} as Record<string, SavedPrice>)

        const { results: docsWithRevision } = await core.dailyPricesDB.bulkGet({ docs: docIds })

        const docs: SavedPrice[] = docsWithRevision.map((x) => ({
          ...documentMap[x.id],
          _rev: "ok" in x.docs[0] ? x.docs[0].ok._rev : undefined,
        }))

        // const dbReceipts =
        await core.dailyPricesDB.bulkDocs(docs)
        // console.log(
        //   "ðŸ“œ LOG > dbReceipts:",
        //   dbReceipts.find((x) => !x.rev?.includes("1-"))
        // )

        const start = (priceMapper(results[0]).time as number) * 1000
        const end = (priceMapper(results[results.length - 1]).time as number) * 1000

        progress([
          (i * 100) / symbols.length,
          `Fetched ${symbol} from ${formatDate(start)} to ${formatDate(end)}`,
        ])

        if (results.length !== PRICE_API_PAGINATION) {
          // reached listing date (genesis)
          break
        }

        until = start - 86400000
        since = start - 86400000 * PRICE_API_PAGINATION
      }
    } catch (error) {
      progress([(i * 100) / symbols.length, `Skipped ${symbol}: ${error}`])
    }
  }
}
