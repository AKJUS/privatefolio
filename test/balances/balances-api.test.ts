import fs from "fs"
import { join } from "path"
import { findAuditLogs } from "src/api/account/audit-logs-api"
import {
  computeBalances,
  getHistoricalBalances,
  getLatestBalances,
} from "src/api/account/balances-api"
import { addFileImport } from "src/api/account/file-imports/file-imports-api"
import { getValue } from "src/api/account/kv-api"
import { fetchDailyPrices } from "src/api/core/daily-prices-api"
import { resetAccount } from "src/api/database"
import { Timestamp } from "src/interfaces"
import { ProgressUpdate } from "src/stores/task-store"
import { beforeAll, expect, it } from "vitest"

const accountName = "blue"

beforeAll(async () => {
  //
  await resetAccount(accountName)
  //
  const fileName = "coinmama.csv"
  const filePath = join("test/files", fileName)
  const buffer = await fs.promises.readFile(filePath, "utf8")
  const file = new File([buffer], fileName, { lastModified: 0, type: "text/csv" })
  await addFileImport(file, undefined, accountName)
})

it.sequential("should not have balances computed", async () => {
  // act
  const auditLogs = await findAuditLogs({}, accountName)
  // assert
  expect(auditLogs).toMatchSnapshot()
})

it.sequential("should compute historical balances", async () => {
  // arrange
  const until = Date.UTC(2017, 8, 14, 0, 0, 0, 0) // 14 Sep 2017
  // act
  const updates: ProgressUpdate[] = []
  await computeBalances((state) => updates.push(state), undefined, {
    accountName,
    until,
  })
  // assert
  const balancesCursor = await getValue<Timestamp>("balancesCursor", undefined, accountName)
  expect(balancesCursor).toBe(until)
  expect(updates.join("\n")).toMatchInlineSnapshot(`
    "0,Computing balances from 2 audit logs
    0,Processing logs 1 to 2
    90,Processing logs 1 to 2 complete
    95,Filling the balances to reach today
    100,Computed all balances!"
  `)
})

it.sequential("should fetch historical balances", async () => {
  // arrange
  // act
  const balances = await getHistoricalBalances({ accountName })
  // assert
  expect(balances.length).toMatchInlineSnapshot(`14`)
  expect(balances).toMatchSnapshot()
})

it.sequential("should have balances computed", async () => {
  // act
  const auditLogs = await findAuditLogs({}, accountName)
  // assert
  expect(auditLogs).toMatchSnapshot()
})

it.sequential("should fetch latest balances without price data", async () => {
  // arrange
  // act
  const balances = await getLatestBalances(accountName)
  // assert
  expect(balances).toMatchSnapshot()
})

it.sequential("should fetch latest balances with price data", async () => {
  // arrange
  await fetchDailyPrices({ symbols: ["BTC"] })
  // act
  const balances = await getLatestBalances(accountName)
  // assert
  expect(balances).toMatchSnapshot()
})
