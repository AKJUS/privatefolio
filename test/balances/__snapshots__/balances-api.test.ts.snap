// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should fetch historical balances 2`] = `
[
  {
    "BTC": 0.035979259,
    "_id": "1498867200000",
    "_rev": "1-15c81300c0d8e274ef580953f32762ce",
    "timestamp": 1498867200000,
  },
  {
    "BTC": 0.035979259,
    "_id": "1498953600000",
    "_rev": "1-34e71d07484303266cc51afd68db3a8d",
    "timestamp": 1498953600000,
  },
  {
    "BTC": 0.035979259,
    "_id": "1499040000000",
    "_rev": "1-4f244f5474c20fd237fab3d898a8ba3d",
    "timestamp": 1499040000000,
  },
  {
    "BTC": 0.035979259,
    "_id": "1499126400000",
    "_rev": "1-d0ad91addd89dfd2ea85084886ef31a4",
    "timestamp": 1499126400000,
  },
  {
    "BTC": 0.035979259,
    "_id": "1499212800000",
    "_rev": "1-a3c853bac5dfbf203d2f554decf9e925",
    "timestamp": 1499212800000,
  },
  {
    "BTC": 0.035979259,
    "_id": "1499299200000",
    "_rev": "1-65c4bd62bc2d54df21aabebe517ce56b",
    "timestamp": 1499299200000,
  },
  {
    "BTC": 0.072958518,
    "_id": "1499385600000",
    "_rev": "2-e6c86db56d5b9f4c14315090a485ba9d",
    "timestamp": 1499385600000,
  },
  {
    "BTC": 0.072958518,
    "_id": "1499472000000",
    "_rev": "1-80a0bf987ef43e4b9783b6fbc99d294a",
    "timestamp": 1499472000000,
  },
  {
    "BTC": 0.072958518,
    "_id": "1499558400000",
    "_rev": "1-75a20de6252314c2908e75661bc72b34",
    "timestamp": 1499558400000,
  },
  {
    "BTC": 0.072958518,
    "_id": "1499644800000",
    "_rev": "1-d1690d8220f67168df907e3d79a4ecfc",
    "timestamp": 1499644800000,
  },
  {
    "BTC": 0.072958518,
    "_id": "1499731200000",
    "_rev": "1-a1a858bf861a17be3cfddaa9ed80179c",
    "timestamp": 1499731200000,
  },
  {
    "BTC": 0.072958518,
    "_id": "1499817600000",
    "_rev": "1-a8dd38c7db98ef6143d236868670d58a",
    "timestamp": 1499817600000,
  },
  {
    "BTC": 0.072958518,
    "_id": "1499904000000",
    "_rev": "1-5951b6d3def143470edfc758a669a8fa",
    "timestamp": 1499904000000,
  },
  {
    "BTC": 0.072958518,
    "_id": "1499990400000",
    "_rev": "1-0a1ff762b5e924bc9bf945f9ae815a6f",
    "timestamp": 1499990400000,
  },
]
`;

exports[`should fetch latest balances 1`] = `
[
  {
    "_id": "1499990400000_BTC",
    "balance": 0.072958518,
    "price": undefined,
    "symbol": "BTC",
    "value": undefined,
  },
]
`;

exports[`should not have balances computed 1`] = `
[
  {
    "_id": "1217276216_2903060452_0",
    "_rev": "1-9563eb7da39f8735f46ea11fbcf9fd55",
    "change": "0.036979259",
    "changeN": 0.036979259,
    "integration": "coinmama",
    "operation": "Buy with Credit Card",
    "symbol": "BTC",
    "timestamp": 1499433076000,
    "wallet": "Spot",
  },
  {
    "_id": "1217276216_1332563442_0",
    "_rev": "1-b5576d4e5c403eced534343ae7af054b",
    "change": "0.035979259",
    "changeN": 0.035979259,
    "integration": "coinmama",
    "operation": "Buy with Credit Card",
    "symbol": "BTC",
    "timestamp": 1498911076000,
    "wallet": "Spot",
  },
]
`;
